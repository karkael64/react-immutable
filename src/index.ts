export { useBase } from "./useBase";
export { useCallbackBase } from "./useCallbackBase";
export { useImmutable } from "./useImmutable";
export { useMemoBase } from "./useMemoBase";
export { useEntries } from "./useEntries";
export { useWritable } from "./useWritable";

export { Readable, readable } from "./readable";
export { Writable, writable } from "./writable";

export type { EntriesMethods } from "./useEntries";
export type { ReadableCallback, ReadableListener } from "./readable";
export type {
  WritableListener,
  WritableReducer,
  WritableUpdater,
} from "./writable";
