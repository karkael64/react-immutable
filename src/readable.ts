import { Writable, WritableListener } from "./writable";

export type ReadableListener<State> = WritableListener<State>;
type ReadableUpdater<State> = (newValue: State) => void;
export type ReadableCallback<State> = (
  updater: ReadableUpdater<State>
) => undefined | (() => void);
type ReadableUnsubscribe = () => void;

export class Readable<State> {
  private value: State;

  private listeners: ReadableListener<State>[] = [];

  private unsubscribeCb?: ReadableUnsubscribe;

  constructor(reader: ReadableCallback<State>) {
    const updater = (newValue: State) => this.set(newValue);
    this.unsubscribeCb = reader(updater);
  }

  /**
   * add a function listening this Writable changes (update or set)
   * @param {ReadableListener<State>} listener function executed at each changes (update or set) of this Writable, executed immediatly.
   */
  subscribe(listener: ReadableListener<State>) {
    this.listeners.push(listener);
    listener(this.value);

    const unsubscribe = () => {
      const index = this.listeners.indexOf(listener);
      if (index !== 1) {
        this.listeners.splice(index, 1);
        return true;
      }
      return false;
    };

    return unsubscribe;
  }

  unsubscribe() {
    if (this.unsubscribeCb) this.unsubscribeCb?.();
    else throw new Error("This Readable instance has no unsubscribe callback");
  }

  private set(value: State) {
    if (value !== this.value) {
      this.value = value;
      for (const listener of this.listeners) {
        listener(this.value);
      }
    }
  }

  valueOf(): State {
    return this.value;
  }

  toString(): string {
    return `${this.value}`;
  }
}

export const readable = <State>(reader: ReadableCallback<State>) =>
  new Readable(reader);

export const readableFromWritable = <State, Input>(
  writable: Writable<State, Input>
) =>
  new Readable<State>((set) =>
    writable.subscribe((newValue) => {
      set(newValue);
    })
  );
