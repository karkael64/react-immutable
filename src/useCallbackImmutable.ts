import { useRef } from "react";

/**
 * this hook creates an immutable callback, even if `fn` changes. It is same as `useCallback`, but does not need dependencies.
 */
export const useCallbackImmutable = <T extends (...args: any[]) => any>(
  fn: T
): T => {
  if (typeof fn !== "function") {
    throw new Error("Parameter should be a function");
  }

  const ref = useRef({
    callback: fn,
    immutable: (...args: Parameters<T>): ReturnType<T> =>
      ref.current.callback(...args),
  });

  Object.assign(ref.current, { callback: fn });
  return ref.current.immutable as T;
};
