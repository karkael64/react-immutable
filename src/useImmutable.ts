import { useState } from "react";

/**
 * this hook instanciates a variable which is immutable in React life cycle.
 * @warn if you destroy the component, this hook will be trigger for the component replacing it
 * @param definitiveState the definitive state of your hook, may be retruned by a function
 * @returns same value as `definitiveState` (if it is a function, returns the returned value).
 * @example
 * ```ts
 * const Comp: React.FC<{ initFormData: Object }> = ({ initFormData }) => {
 *   const savedInitFormData = useImmutable(initFormData);
 *   return (<>jsx</>);
 * };
 * ```
 * @example
 * YOU SHOULD NOT CODE LIKE THIS
 * ```ts
 * const Comp: React.FC = () => {
 *   const text = useImmutable('Hello world!');
 *   return (<p>{text}</p>);
 * };
 * ```
 * PLEASE USE THIS CODE INSTEAD
 * ```ts
 * const text = 'Hello world!';
 * const Comp: React.FC = () => {
 *   return (<p>{text}</p>);
 * };
 * ```
 */
export const useImmutable = <T>(definitiveState: (() => T) | T): T =>
  useState(definitiveState)[0];
