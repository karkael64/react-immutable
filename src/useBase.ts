import { useImmutable } from "./useImmutable";

/**
 * this hook writes properties of `entry` in an immutable (in React life cycle) object `base`, which is usefull for referencing items that should be read only.
 * @param entry is any object (the prototype and not enumerable properties of `entry` are ignored)
 * @returns a new object immutable in React life cycle, with properties copied from `entry`, but without object prototype
 * @example
 * remember an immutable is useless in a useEffect dependencies: in this code below useCallback dependencies `[base, setCount]` are equivalent to `[]`, which means callback returned is also immutable.
 * ```ts
 * const Comp: React.FC = () => {
 *   const [count, setCount] = useState(0);
 *   const base = useBase({ count });
 *   // `base` and `setCount` are immutable, so `handleClick` is immutable here too
 *   const handleClick = useCallback(() => setCount(base.count + 1), [base, setCount]);
 *   return <button onClick={handleClick}>{count}</button>;
 * };
 * ```
 * @example
 * ```ts
 * const Comp: React.FC({ count: number; onClick(newValue: number): void }) = ({Â count, onClick }) => {
 *   const [count, setCount] = useState(0);
 *   const base = useBase({ count });
 *   const handleClick = useImmutable(() => () => onClick(base.count + 1));
 *   return <button onClick={handleClick}>{count}</button>;
 * };
 * ```
 */
export const useBase = <
  T extends Record<string, any>,
  R extends Record<string, any> = { [k in keyof T]: T[k] }
>(
  entry: T
): R => {
  if (!(typeof entry === "object"))
    throw new Error("First parameter should be an object");
  const base = useImmutable({} as R);
  return Object.assign(base, entry);
};
