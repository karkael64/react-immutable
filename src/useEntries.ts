import { useImmutable } from "./useImmutable";

export type EntriesMethods<T> = {
  getFirst: () => T | undefined;
  getLast: () => T | undefined;
  getEntries: () => T[];
  getUnique: () => T[];
  getChangingList: () => T[];
  countEntries: () => number;
};

/**
 * this hook register a state entry each React render life cycle
 * @param entry state registered
 * @returns an immutable object with methods to get entries
 * @example
 * ```ts
 * const Comp: React.FC<> = () => {
  const [state, setState] = useState(-4.2);
  const entries = useEntries(state);
  const handleReset = useImmutable(() => setState(entries.getFirst()));

  return <Form state={state} setState={setState} onReset={handleReset} />;
};
 * ```
 */

export const useEntries = <T>(entry: T) => {
  const [list, methods] = useImmutable((): [T[], EntriesMethods<T>] => [
    [],
    {
      getFirst: () => list[0],
      getLast: () => list[list.length - 1],
      getEntries: () => list.slice(),
      getUnique: () =>
        Array.from(
          list.reduce((acc, item) => {
            acc.add(item);
            return acc;
          }, new Set<T>())
        ),
      getChangingList: () =>
        list.reduce((acc, item) => {
          if (acc[0] === undefined || acc[0] !== item) {
            acc.push(item);
          }
          return acc;
        }, [] as T[]),
      countEntries: () => list.length,
    },
  ]);

  list.push(entry);

  return methods;
};
