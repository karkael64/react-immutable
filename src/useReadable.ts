import { useEffect, useState } from "react";
import { Readable, ReadableCallback, isReadable, readable } from "./readable";
import { useImmutable } from "./useImmutable";

type UseReadableOuput<State> = [State, Readable<State>];
type UseReadable = {
  <State>(reader: Readable<State>): UseReadableOuput<State>;
  <State>(reader: ReadableCallback<State>): UseReadableOuput<State>;
};

/**
 * @function `useReadable` returns a React state management of a Readable
 * @param {ReadableCallback<State>} reader is an optional function for reducing new value at update, used only if `init` is not a Readable
 * @returns {UseReadableOuput<State>} is a constant (do not change in React LifeCycle) list of arguments, where first item is the current value of Readable, second item is a constant updater of Readable, third item is a constant Readable
 */
export const useReadable: UseReadable = <State>(
  reader: Readable<State> | ReadableCallback<State>
): UseReadableOuput<State> => {
  const [state, setState] = useState<State | undefined>(
    isReadable(reader) ? reader.valueOf() : undefined
  );

  const unmutable = useImmutable((): UseReadableOuput<State> => {
    const binded = isReadable(reader) ? reader : readable(reader);
    return [binded.valueOf(), binded];
  });

  useEffect(() => {
    const unsubscribe = unmutable[1].subscribe(setState);
    return () => {
      unsubscribe();
    };
  }, []);

  unmutable[0] = state as State;
  return unmutable;
};
