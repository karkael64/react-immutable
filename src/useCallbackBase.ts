import { useBase } from "./useBase";
import { useImmutable } from "./useImmutable";

/**
 * this hook creates an immutable callback, with component states registered in `values` available
 * as prepend to others
 * @example
 * ```ts
 * const Comp: React.FC<{ id: number }> = ({ id }) => {
 *   const [count, setCount] = useState(0);
 *   const handleClick = useCallbackBase({ id, count }, (base) => () => {
 *     console.log('try to call', base.id);
 *     setCount(base.count + 1);
 *   });
 *   return (<button onClick={handleClick}>{count}</button>)
 * };
 * ```
 * @example
 * YOU SHOULD NOT CODE LIKE THIS: each time this component or a parent component refresh it life
 * cycle (a state changed), `handleClick` is a new function, then React unassign previous
 * `handleClick` of button and assign the new `handleClick` on click event.
 * ```ts
 * const Comp: React.FC<{ id: number }> = ({ id }) => {
 *   const handleClick = () => {
 *     console.log('try to call', id);
 *   };
 *   return (<button onClick={handleClick}>{id}</button>)
 * };
 * ```
 * YOU SHOULD NOT CODE LIKE THIS: each time `count` changes, the function `handleClick` changes
 * too. It also means React unassign previous `handleClick` of button and assign the new
 * `handleClick` on click event.
 * ```ts
 * const Comp: React.FC<{ count: number }> = ({ count }) => {
 *   const handleClick = useCallback(() => {
 *     console.log('call count', count);
 *   }, [id]);
 *   return (<button onClick={handleClick}>{count}</button>)
 * };
 * ```
 * PLEASE USE THIS CODE INSTEAD: `handleClick` is immutable (in React life cycle) and read component
 * state passed by reference `base`.
 * ```ts
 * const Comp: React.FC<{ id: number, count: number }> = ({ id, count }) => {
 *   const handleClick = useCallbackBase({ id, count }, (base) => () => {
 *     console.log('try to call', base.id);
 *     console.log('call count', base.count);
 *   });
 *   return (<button onClick={handleClick}>{id}</button>)
 * };
 * ```
 */
export const useCallbackBase = <
  Values extends Record<string, any>,
  Fn extends (base: Values) => (...args: any[]) => any
>(
  values: Values,
  fn: Fn
) => {
  if (!(typeof fn === "function"))
    throw new Error("First parameter should be a function");
  const base = useBase<Values>(values);
  return useImmutable(() => fn(base) as ReturnType<Fn>);
};
