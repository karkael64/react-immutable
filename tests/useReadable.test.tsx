import { render, waitFor } from "@testing-library/react";
import { useEffect } from "react";
import { expect, it, test } from "vitest";
import { useReadable } from "../src/useReadable";
import { Readable, readableFromWritable } from "../src/readable";
import "global-jsdom/register";
import { Writable, writable } from "../src";

// global usage
const textStorage = writable("");
const textToWrite = "Hello Wesley";
const text = readableFromWritable(textStorage);
const countStorage = writable(0);

const CompGlobalA: React.FC = () => {
  const [textState] = useReadable(text);

  useEffect(() => {
    if (textState.length < textToWrite.length && textState.length % 2) {
      setTimeout(() =>
        textStorage.update((current, set) =>
          set(textState + textToWrite[textState.length])
        )
      );
    }
  }, [textState]);

  return <span>{textState}</span>;
};

const CompGlobalB: React.FC = () => {
  const [textState] = useReadable(text);

  useEffect(() => {
    if (textState.length < textToWrite.length && (textState.length + 1) % 2) {
      setTimeout(() =>
        textStorage.set(textState + textToWrite[textState.length])
      );
    }
  }, [textState]);

  return <span>{`${text}`}</span>;
};

const CompNotListening: React.FC = () => {
  return <span>{`${text}`}</span>;
};

// scoped usage
const CompInner: React.FC<{ count: Readable<number | string> }> = ({
  count,
}) => {
  const [countState] = useReadable(count);

  return <span>{`${countState}`}</span>;
};

const CompOuter: React.FC<{ countStorage: Writable<number> }> = ({
  countStorage,
}) => {
  const [countState, countBind] = useReadable<number>((set) =>
    countStorage.subscribe(set)
  );
  setTimeout(() => countStorage.set(3));

  return (
    <div>
      <CompInner count={countBind} />
      <span>{countState}</span>
      <span>{typeof countState}</span>
    </div>
  );
};

it("should use bind globally", async () => {
  const { container, getAllByText } = render(
    <div>
      <CompGlobalA />
      <CompGlobalB />
      <CompNotListening />
    </div>
  );

  expect(container).toMatchSnapshot();
  await waitFor(() => expect(getAllByText(textToWrite)).toHaveLength(2));
  expect(container).toMatchSnapshot();
});

it("should use bind on properties", async () => {
  const { container, getAllByText } = render(
    <CompOuter countStorage={countStorage} />
  );

  expect(container).toMatchSnapshot();
  await waitFor(() => expect(getAllByText("3")).toHaveLength(2));
  expect(container).toMatchSnapshot();
});
