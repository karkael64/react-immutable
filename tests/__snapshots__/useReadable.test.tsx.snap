// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should use bind globally 1`] = `
<div>
  <div>
    <span />
    <span />
    <span />
  </div>
</div>
`;

exports[`should use bind globally 2`] = `
<div>
  <div>
    <span>
      Hello Wesley
    </span>
    <span>
      Hello Wesley
    </span>
    <span />
  </div>
</div>
`;

exports[`should use bind on properties 1`] = `
<div>
  <div>
    <span>
      0
    </span>
    <span>
      0
    </span>
    <span>
      number
    </span>
  </div>
</div>
`;

exports[`should use bind on properties 2`] = `
<div>
  <div>
    <span>
      3
    </span>
    <span>
      3
    </span>
    <span>
      number
    </span>
  </div>
</div>
`;
